### DISCUSSION TITLE
Excessive memory usage

### DISCUSSION THREAD

**Discussion by antoinecarme** [duckdb/duckdb#5075] [General]
My journey with duckdb continues ... 

I have a SQL code that consumes a lot of memory when executing (compared to PostgreSQL, SQLite). This code is generated automatically from a machine learning model and reflects the architecture of the model itself (random forest with 32 trees). The SQL generation method has been tested on 10 different database dialects and is OK (produces the right result in a reasonable time).

Don't know if this is really a bug, so I am starting a discussion here.

This is a "large" SQL code, that contains 32 groups of CTEs with internal aggregates. A final select statement computes the means of individual columns from all the CTEs. 

All the joins used are "left outer". So the 32 CTEs have same number of rows rows each as the input table (digits)

Executing this SQL code fails when setting the memory limit to 1GB and is OK when using 10GB.

Is there a way to make the optimizer handle this code more efficiently by using less memory?

To reproduce using duckdb CLI :

```
# wget -O digits.csv https://raw.githubusercontent.com/antoinecarme/sklearn2sql_heroku/master/docs/data/digits.csv
# wget -O 32.sql https://raw.githubusercontent.com/antoinecarme/sklearn2sql_heroku/master/docs/data/digits_64_features_RF_32_DuckDB.sql

# Even with 2 rows, the problem can be reproduced.
CREATE TABLE INPUT_DATA AS SELECT * FROM 'digits.csv' LIMIT 2;
SET  memory_limit="1000mb";
.read '32.sql'
```

As this is an out of memory issue, I suppose that there are some flags to make the optimizer handle this code and generate a more efficient plan. 

Similar/Larger SQL codes for larger models (512 trees) are available in the same directory and are OK on PostgreSQL and SQLite.

Thanks in advance

Antoine

**Comment by antoinecarme** [comment]
For what it is worth, a gdb session shows that the out of memory error is linked with some functions related to aggregates and perfect hashing.

**Comment by Mytherin** [comment]
DuckDB is intended to process larger amounts of data. Internally it allocates buffers on the order of hundreds of kilobytes~a few megabytes per operator per thread. That means that if you have a complex query tree, there will be a minimum amount of RAM required to process that query tree just to allocate these buffers, even if you are only pushing a few tuples through it. Postgres and SQLite by contrast push individual tuples through a pipeline and therefore require less memory when working with only a few tuples. 

This query tree just seems to be very complex. Perhaps try splitting the data generation into separate statements and materialize that into temporary tables?

**Comment by antoinecarme** [comment]
@Mytherin 


Thanks a lot for your quick answer.

My use case is typically computational (use the database as a cpu/gpu and compute scores/probabilities for one tuple at a time). There is no need for storage.

Is there a way to control the size of the memory per operator / per thread ? just to make it behave like sqlite or postgresql ?

**Reply by Mytherin** [reply]
That does not exist right now, no. In general if your use case is processing a few tuples I would say DuckDB is not what you are looking for. Outside of memory concerns, it is also just not optimized for it. The entire system is designed around batch processing of tens of thousands of tuples.

**Comment by antoinecarme** [comment]
@Mytherin 

Materialization of intermediate CTEs helps a lot in this case. SQL code execution is extremely fast now even for the large model (512) case. 

https://github.com/antoinecarme/sklearn2sql_heroku/blob/master/docs/data/digits_64_features_RF_32_DuckDB_with_materialization.sql

Thanks a lot for the hint. 

Problem Solved.

**Reply by Alex-Monahan** [reply]
Nice work!! Do you happen to have any benchmark results you could share? Thanks!
